src/main/java/tasks/Task.java

```diff
@@ -0,0 +1,119 @@
+package tasks;
+
+import duke.DukeDate;
+
+/**
+ * A class that abstracts a task.
+ */
+public abstract class Task {
+
+    /**
+     * A representation of the 3 types of tasks currently supported by Duke.
+     */
+    public enum Type {
+        TODO,
+        DEADLINE,
+        EVENT
+    }
+
+    /** Whether the task had been completed or not */
+    private boolean isDone = false;
+    /** The name of a task as given by the user */
+    private final String taskName;
+    /** The type of the task */
+    private final Type type;
+
+
+    protected Task(String taskName, Type type) {
+        this.taskName = taskName;
+        this.type = type;
+    }
+
+    /**
+     * Returns a description of the task filled with its details.
+     *
+     * @return The description of the task.
+     */
+    public abstract String taskDescription();
+
+    /**
+     * Converts a task into a specified String format for saving into local storage.
+     * The format to save the strings is {type of task} | {is done} | {type of date} |
+     * {task name} | {time (if applicable)}.
+     * The type of date is based on the static variables in the duke.DukeDate class. It is only applicable
+     * for deadline and event tasks. For other tasks with no date, a "0" is used to represent the type
+     * of date.
+     *
+     * @return The String format to save the task as.
+     */
+    public abstract String taskSaveString();
+
+    /**
+     * Marks a task as completed.
+     */
+    public void setAsFinished() {
+        this.isDone = true;
+    }
+
+    /**
+     * Returns the string representation of the type of task.
+     *
+     * @return The required string representation.
+     */
+    public String eventTypeToString() {
+        if (this.type.equals(Type.TODO)) {
+            return "[T]";
+        } else if (this.type.equals(Type.EVENT)) {
+            return "[E]";
+        } else {
+            return "[D]";
+        }
+    }
+
+    /**
+     * Returns a task from the String used to saved the task in local storage. When a task is saved to local
+     * storage it is saved as a formatted string that stores the information of the task. This methods takes
+     * that String and converts it back to the correct task.
+     *
+     * @param saveString The String that represents the saved task.
+     * @return The task converted from saveString.
+     */
+    public static Task stringToTask(String saveString) {
```

Comment:
Maybe this method could be renamed to something like storageStringToTask, just to make it clear that its specifically talking about parsing the storageString format